backends/bmv2/common/parser.cpp

```diff
@@ -443,7 +443,8 @@ bool ParserConverter::preorder(const IR::P4Parser* parser) {
 
     for (auto s : parser->parserLocals) {
         if (auto inst = s->to<IR::P4ValueSet>()) {
-            auto bitwidth = inst->elementType->width_bits();
+            auto type = ctxt->typeMap->getTypeType(inst->elementType, true);
```

Comment:
I think it is better to ensure that elementType expression has the proper type assigned.

Reply:
I did look at the typeChecker code which canonicalizes elementType as shown below.

https://github.com/p4lang/p4c/blob/master/frontends/p4/typeChecking/typeChecker.cpp#L1256

The output from typeChecker has elementType as Type_Name which looks correct to me.   The backend needs  to use getTypeType() if type is Type_Name.

Reply:
I debugged `const IR::Node* TypeInference::postorder(IR::P4ValueSet* decl)` and see an issue.  The `decl` coming in and being returned does not change `decl->elementType`.  The coming in type for `decl->elementType`  in `Type_Name`.  I changed the frontend code, backed out the bmv2 backend code and the issue is fixed.  I will update the code changes shortly.

Reply:
It will be too complicated to change frontend typeChecker for P4ValueSet to set the elementType to a proper type.  The reason is because the postorder(IR::P4ValueSet) method is entered multiple times and additional checks are needed to not create new node with proper element type. 

This is why you may note, Mihai copied the backend changes of this PR and issued another PR to fix the issue in the backend as I proposed here.